types/types.cc

```diff
@@ -3679,11 +3679,17 @@ make_user_value(data_type type, user_type_impl::native_type value) {
     return data_value::make_new(std::move(type), std::move(value));
 }
 
-std::ostream& operator<<(std::ostream& out, const data_value& v) {
+ auto fmt::formatter<data_value>::format(const data_value& v,
+                                         fmt::format_context& ctx) const -> decltype(ctx.out()) {
     if (v.is_null()) {
-        return out << "null";
+        return fmt::format_to(ctx.out(), "null");
     }
-    return out << v.type()->to_string_impl(v);
+    return fmt::format_to(ctx.out(), "{}", v.type()->to_string_impl(v));
```

Comment:
Unfortunately I'm quite clueless on how fmt works under the hood. Is this format_to with a fixed format string "{}" as efficient as a string copy? Does the parsing of the silly string "{}" happen in compile time (how?) or runtime (inefficient...)?

Reply:
> Unfortunately I'm quite clueless on how fmt works under the hood. Is this format_to with a fixed format string "{}" as efficient as a string copy? 

it's unfair to compare it with a plain `strcpy()`, as the `fmt::formatter` default-generated from `operator<<` always creates yet another `basic_memory_buffer<char>` in its own `format(const T& value, basic_format_context<OutputIt, Char>& ctx)`. after this change, the customized formatter just print into the provided `ctx.out()`, without the help of the `basic_memory_buffer<char>`.

> Does the parsing of the silly string "{}" happen in compile time (how?) 

it happens in compile-time if the compile-time format string is used as in this use case. something like
```py
def format(format_string, *args):
    buffer = memory_buffer()
    fmt_ctx = buffer.begin()
    begin, end = 0, len(format_string)
    arg = iter(args)
    while begin != end:
        p = format_string.find('{', begin)
        if p >= 0:
            fmt_ctx = copy(fmt_ctx, format_string[begin:p])
            p += 1
            formatter = fmt.formatter(type(next(arg))
            if format_string[p] == ':':
                format_specifier = format_string[p:]
                p = formatter.parse(format_specifier)
                fmt_ctx = formatter.format(arg, fmt_ctx)
            else:
                p += 1
            assert format_string[p] == '}', "missing '}' in format string"
            begin = p + 1
        else:
            fmt_ctx = copy(fmt_ctx, format_string[begin:])
            begin = end
    return buffer
```

fmt uses a dedicated class named `format_string_checker` to perform the output for compile-time checking.

> or runtime (inefficient...)?

and it also happens in runtime. if well, this piece of code is executed.